viewer/src/utilities/IndexSet.ts

```diff
@@ -0,0 +1,594 @@
+      if (originalRanges[i].toInclusive + 1 >= originalRanges[i + 1].from) {
+        // Should not happen, but let's be safe
+        continue;
+      }
+      newRanges.push(NumericRange.createFromInterval(originalRanges[i].toInclusive + 1, originalRanges[i + 1].from));
+    }
+
+    return newRanges;
+  }
+
+  unionWith(otherSet: IndexSet): IndexSet {
+    if (this.rootNode) {
+      const rs = otherSet.ranges();
+      rs.forEach(range => {
+        this.rootNode = (this.rootNode as IndexNode).addRange(range);
+      });
+    } else {
+      this.rootNode = otherSet.rootNode;
+    }
+
+    return this;
+  }
+
+  differenceWith(otherSet: IndexSet): IndexSet {
+    if (this.rootNode) {
+      const rs = otherSet.ranges();
+      rs.forEach(range => {
+        if (this.rootNode) {
+          this.rootNode = (this.rootNode as IndexNode).removeRange(range);
+        }
+      });
+    }
+
+    return this;
+  }
+
+  hasIntersectionWith(otherSet: IndexSet | Set<number>): boolean {
+    if (otherSet instanceof IndexSet) {
+      if (this.rootNode == undefined || otherSet.rootNode == undefined) {
+        return false;
+      }
+
+      return this.rootNode.hasIntersectionWith(otherSet.rootNode);
+    } else {
+      for (const index of otherSet) {
+        if (this.contains(index)) {
+          return true;
+        }
+      }
+
+      return false;
+    }
+  }
+
+  intersectWith(otherSet: IndexSet): IndexSet {
+    if (this.rootNode && otherSet.rootNode) {
+      // Tackle endpoints
+      // Remove left bounds outside input node set
+      if (this.rootNode.range.from < otherSet.rootNode.range.from) {
+        const leftBoundRange = NumericRange.createFromInterval(
+          this.rootNode.range.from,
+          otherSet.rootNode.range.from - 1
+        );
+        this.rootNode = this.rootNode.removeRange(leftBoundRange);
+
+        if (!this.rootNode) {
+          return this;
+        }
+      }
+
+      // Remove right bounds outside input node set
+      if (this.rootNode.range.toInclusive > otherSet.rootNode.range.toInclusive) {
+        const rightBoundRange = NumericRange.createFromInterval(
+          otherSet.rootNode.range.toInclusive + 1,
+          this.rootNode.range.toInclusive
+        );
+        this.rootNode = this.rootNode.removeRange(rightBoundRange);
+      }
+
+      // Invert otherSet ranges and remove them
+      const invRanges = otherSet.invertedRanges();
+
+      invRanges.forEach(range => {
+        if (this.rootNode) {
+          this.rootNode = this.rootNode.removeRange(range);
+        }
+      });
+    } else if (this.rootNode) {
+      // Otherset is empty, set this to empty as well
+      this.rootNode = undefined;
+    }
+    return this;
+  }
+
+  clear(): void {
+    this.rootNode = undefined;
+  }
+
+  clone(): IndexSet {
+    const st: IndexSet = new IndexSet();
+
+    if (this.rootNode) {
+      st.rootNode = this.rootNode.clone();
+    }
+
+    return st;
+  }
+}
```

Comment:
Anyway we can logically split this up with helper classes? 300+ lines file feels bad, we have enough of those already 😂.

Reply:
Fixed